File: client/src/Models/ServerMessages.ts
import { OrderStatus, ServerMessageType } from "types/Enums"
import { Envelope, Instrument, Message, Quote } from "./Base"

export type ServerEnvelope =
	| ErrorInfoEnvelope
	| SuccessInfoEnvelope
	| ExecutionReportEnvelope
	| MarketDataUpdateEnvelope

export interface ServerMessage extends Message {}

export interface ErrorInfo extends ServerMessage {
	reason: string
}

export interface ErrorInfoEnvelope extends Envelope {
	messageType: ServerMessageType.error
	message: ErrorInfo
}

export interface SuccessInfo extends ServerMessage {
    info: any
}

export interface SuccessInfoEnvelope extends Envelope {
	messageType: ServerMessageType.success
	message: SuccessInfo
}

export interface ExecutionReport extends ServerMessage {
	orderId: string
	orderStatus: OrderStatus
}

export interface ExecutionReportEnvelope extends Envelope {
	messageType: ServerMessageType.executionReport
	message: ExecutionReport
}

export interface MarketDataUpdate extends ServerMessage {
	subscriptionId: number
	instrument: Instrument
	quotes: [Quote]
}

export interface MarketDataUpdateEnvelope extends Envelope {
	messageType: ServerMessageType.marketDataUpdate
	message: MarketDataUpdate
}

File: server/app/schemas/server_messages.py
from __future__ import annotations

from typing import List, TypeVar

import bidict as bidict

from app.utils import enums
from app.schemas.base import Envelope, InstrumentData, Message, QuoteData


class ServerMessage(Message):
    def get_type(self: ServerMessageT) -> enums.ServerMessageType:
        return _SERVER_MESSAGE_TYPE_BY_CLASS[self.__class__]


class ErrorInfo(ServerMessage):
    reason: str


class SuccessInfo(ServerMessage):
    info: str | dict


class ExecutionReport(ServerMessage):
    order_id: str
    order_status: enums.OrderStatus


class MarketDataUpdate(ServerMessage):
    subscription_id: int
    instrument: InstrumentData
    quotes: List[QuoteData]


class ServerEnvelope(Envelope):
    message_type: enums.ServerMessageType

    def get_parsed_message(self):
        return _SERVER_MESSAGE_TYPE_BY_CLASS.inverse[self.message_type].parse_obj(self.message)


_SERVER_MESSAGE_TYPE_BY_CLASS = bidict.bidict({
    SuccessInfo: enums.ServerMessageType.success,
    ErrorInfo: enums.ServerMessageType.error,
    ExecutionReport: enums.ServerMessageType.execution_report,
    MarketDataUpdate: enums.ServerMessageType.market_data_update,
})
ServerMessageT = TypeVar('ServerMessageT', bound=ServerMessage)
